src/iov.c

```diff
@@ -73,14 +76,20 @@ void ofi_consume_iov_desc(struct iovec *iov, void **desc,
 			break;
 		to_consume -= iov[i].iov_len;
 	}
-	memmove(iov, &iov[i], sizeof(*iov) * (*iov_count - i));
-	if (desc)
-		memmove(desc, &desc[i],
-			sizeof(*desc) * (*iov_count - i));
+	if (i != *iov_count) {
+		memmove(iov, &iov[i], sizeof(*iov) * (*iov_count - i));
+		if (desc)
+			memmove(desc, &desc[i],
+				sizeof(*desc) * (*iov_count - i));
+	}
 	*iov_count -= i;
 out:
-	iov[0].iov_base = (uint8_t *)iov[0].iov_base + to_consume;
-	iov[0].iov_len -= to_consume;
+	if (to_consume < iov[0].iov_len) {
+		iov[0].iov_base = (uint8_t *)iov[0].iov_base + to_consume;
+		iov[0].iov_len -= to_consume;
+	} else {
+		*iov_count -= 1;
```

Comment:
It took me waaaay too much time looking at this function to see if it was returning what we actually want.  The case where we fully consume the iov with an iov_count > 1 is the confusing part.  That flows through the if statement after out: and adjusts iov_base and iov_len by adding/subtracting 0.

In any case, I think we need to make the function more intuitive to understand.  Separating iov_count == 1 into its own section can help.  I also want to avoid repeating the same check twice - once in the for-loop and again after.  This is what I've come up with so far:

```
if (*iov_count == 1) {
    if (to_consume < iov[0].iov_len)
        ofi_consume_sge(iov, to_consume);
    else
        *iov_count = 0;
    return;
}

for (i = 0; i < *iov_count; i++) {
    if (to_consume < iov[0].iov_len) {
        ofi_consume_sge(iov, to_consume)
        break;
    }
    to_consume -= iov[i].iov_len;
}
if (i < *iov_count) {
    memmove(...)
}
iov_count -= i;
```
This removes the -1 adjustment, which now I'm questioning.  Whatever i is after the for loop is the number of iov's that were fully consumed.

Reply:
ofi_consume_sge() would be the iov_base + to_consume / iov_len -= to_consume portion of the original code.  I also noticed that the if statement in my above for loop uses iov[0] instead of iov[i]..

Reply:
And another note, in the code in my comment, we can move the memmove() into the for-loop in the if statement.

Reply:
Apologies Sean, I will be unable to work on this for the next 3 weeks while I'm OOO.

Reply:
Between my first rev and this one you drafted we've pretty substantially re-written the function, so I took a stab at just writing it from scratch.  I liked what I came up with more than your proposed solution.  It has the added benefit that calling consume with iov_count=0 doesn't segfault, never performs a memmove of 0 bytes, and I find it quite readable.  Let me know if it's readable to you as well, or if you can think of another corner case I missed.